stacks-signer/src/chainstate.rs

```diff
@@ -0,0 +1,375 @@
+        );
+
+        let tenures_reorged = client.get_tenure_forking_info(
+            &sortition_state.parent_tenure_id,
+            &sortition_state.prior_sortition,
+        )?;
+        if tenures_reorged.len() == 0 {
+            warn!("Miner is not building off of most recent tenure, but stacks node was unable to return information about the relevant sortitions. Marking miner invalid.";
+                    "proposed_block_consensus_hash" => %block.header.consensus_hash,
+                    "proposed_block_signer_sighash" => %block.header.signer_signature_hash(),
+            );
+            return Ok(false);
+        }
+        for tenure in tenures_reorged.iter() {
+            if tenure.first_block_mined.is_some() {
+                // TODO: must check if the first block was poorly timed.
+                warn!(
+                    "Miner is not building off of most recent tenure, but a tenure they attempted to reorg has already mined blocks.";
+                    "proposed_block_consensus_hash" => %block.header.consensus_hash,
+                    "proposed_block_signer_sighash" => %block.header.signer_signature_hash(),
+                    "parent_tenure" => %sortition_state.parent_tenure_id,
+                    "last_sortition" => %sortition_state.prior_sortition,
+                    "violating_tenure_id" => %tenure.consensus_hash,
+                    "violating_tenure_first_block_id" => ?tenure.first_block_mined,
+                );
+                return Ok(false);
+            }
+        }
+
+        return Ok(true);
+    }
+
+    fn check_tenure_change_block_confirmation(
+        tenure_change: &TenureChangePayload,
+        block: &NakamotoBlock,
+        signer_db: &SignerDb,
+    ) -> Result<bool, ClientError> {
+        // in tenure changes, we need to check:
+        // (1) if the tenure change confirms the expected parent block (i.e.,
+        //     the last block we signed in the parent tenure)
+        // (2) if the parent tenure was a valid choice
+        Self::confirms_known_blocks_in(block, &tenure_change.prev_tenure_consensus_hash, signer_db)
+    }
+
+    fn confirms_known_blocks_in(
+        block: &NakamotoBlock,
+        tenure: &ConsensusHash,
+        signer_db: &SignerDb,
+    ) -> Result<bool, ClientError> {
+        let Some(last_known_block) = signer_db
+            .get_last_signed_block_in_tenure(tenure)
+            .map_err(|e| ClientError::InvalidResponse(e.to_string()))?
+        else {
+            info!(
+                "Have not signed off on any blocks in the parent tenure, assuming block confirmation is correct";
+                "proposed_block_consensus_hash" => %block.header.consensus_hash,
+                "proposed_block_signer_sighash" => %block.header.signer_signature_hash(),
+                "tenure" => %tenure,
+            );
+            return Ok(true);
+        };
+        if block.header.chain_length > last_known_block.block.header.chain_length {
```

Comment:
Do we know whether or not `last_known_block` is an ancestor of `block`?  Because, if there's been a Bitcoin reorg, it might not be.

Reply:
It’s the last known block in the parent tenure of block (identified by consensus hash). A bitcoin fork won’t change this. It’s possible that there’s a mismatch between the parent tenure and the block header’s parent block ID, but that would be an invalid Stacks block, which the block validation endpoint is responsible for rejecting.